runtime/src/commonMain/kotlin/pbandk/impl/Util.kt

```diff
@@ -0,0 +1,157 @@
+            }
+        }
+        return v.toInt() and 0xffff
+    }
+}
+
+private class CodePointIterable (private val s: String): Iterable<Int> {
+    override fun iterator(): Iterator<Int> = CodePointIterator(s)
+}
+
+object Util : UtilInterface {
+    override fun stringToUtf8(str: String): ByteArray {
+        val output = ByteArray(utf8Len(str))
+        var pos = 0
+        CodePointIterator(str).forEach {
+            when (it) {
+                in 0..0x7f -> output[pos++] = it.toByte()
+                in 0x80..0x7ff -> {
+                    output[pos++] = (0xc0 or ((it shr 6) and 0x1f)).toByte()
+                    output[pos++] = (0x80 or (it and 0x3f)).toByte()
+                }
+                in 0x800..0xffff -> {
+                    output[pos++] = (0xe0 or ((it shr 12) and 0xf)).toByte()
+                    output[pos++] = (0x80 or ((it shr 6) and 0x3f)).toByte()
+                    output[pos++] = (0x80 or (it and 0x3f)).toByte()
+                }
+                else -> {
+                    output[pos++] = (0xf0 or ((it shr 18) and 0x7)).toByte()
+                    output[pos++] = (0x80 or ((it shr 12) and 0x3f)).toByte()
+                    output[pos++] = (0x80 or ((it shr 6) and 0x3f)).toByte()
+                    output[pos++] = (0x80 or (it and 0x3f)).toByte()
+                }
+            }
+        }
+        return output
+    }
+    override fun utf8ToString(bytes: ByteArray): String {
+        var pos = 0
+        val ret = StringBuilder()
+        while (pos < bytes.size) {
+            val codePoint: Int
+            when (bytes[pos].toInt() and 0xff) {
+                in 0..0x7f -> {
+                    codePoint = bytes[pos].toInt()
+                    pos++
+                }
+                in 0xc0..0xdf -> {
+                    codePoint = decodeSeq(bytes, pos, 2)
+                    pos += 2
+                }
+                in 0xe0..0xef -> {
+                    codePoint = decodeSeq(bytes, pos, 3)
+                    pos += 3
+                }
+                in 0xf0..0xf7 -> {
+                    codePoint = decodeSeq(bytes, pos, 4)
+                    pos += 4
+                }
+                else -> throw InvalidUTF8Exception()
+            }
+            when (codePoint) {
+                in 0..0xffff -> ret.append(codePoint.toChar())
+                in 0x10000..0x10ffff -> ret.append(
+                        (((codePoint shr 10) and 0x3ff) or 0xDC00).toChar(),
+                        ((codePoint and 0x3ff) or 0xD800).toChar())
+                else -> throw InvalidUTF8Exception()
+            }
+        }
+        return ret.toString()
+    }
+
+    @ExperimentalStdlibApi
+    override fun base64ToBytes(str: String): ByteArray {
+        return str.encodeToByteArray()
+    }
+
+    override fun bytesToBase64(bytes: ByteArray): String {
+        return bytes.contentToString()
+    }
```

Comment:
Maybe not working indeed ... 🤔

Having a hard time with this ... Tried okio's Base64 class which require a lot of code from library. And we can't use Base64 class directly since it's internal. Found another code sample full-ish Kotlin implementation of Base64 [here](https://gist.github.com/hrules6872/e2d4d02a1e8d3c6328ae5aeabc430b96) but `String()` require an array of char and not bytes so not working...

Reply:
I can take a stab at getting something working

Reply:
Drop this `Base64.kt` file into `runtime/src/commonMain/kotlin/pbandk/impl/Base64.kt`: https://gist.github.com/garyp/4860b91d8ae33faf23e081d4290a3532. Then you can use it from `Util` like this:

```kt
    override fun base64ToBytes(str: String): ByteArray {
        return str.decodeBase64ToArray()
            ?: throw RuntimeException("Unable to base64-decode string: $str")
    }

    override fun bytesToBase64(bytes: ByteArray): String {
        return utf8ToString(bytes.encodeBase64())
    }
```